### PR TITLE
Add support for pausing and resuming emulation

### LINKED ISSUE #2193: Potential race condition during Pause/CreateSnapshot sequence
Looking at the implementation of Firecracker's snapshot functionality, the events happen in the following order:
- Pause Vcpus (VMM thread is still running and can process a Block request - for example a write to disk)
- At this point the Firecracker customers can copy the block device files to a new location; This is where the race condition can happen because the write request is still in transit and VMM emulation is still running
- CreateSnapshot

Even when saving the block device happens after the CreateSnapshot API, there is still the opportunity of a race because the Pause and CreateSnapshot requests can be sent back to back via API connection reuse.

Thanks @andreeaflorescu for finding this.

### PR DESCRIPTION
**Pull request by acatangiu**
## Reason for This PR

Fixes #2193 
Alternative to #2208 

## Description of Changes

In order to get the simplest fix in, we addressed this at the API server communication layer (`api_server_adapter.rs`), by blocking the `EventManager` loop in the API event processing function on receiving `VmmAction::Pause`.

The API event processing function then runs in a loop, synchronously processing further API events until `VmmAction::Resume` when things go back to normal.

Because we don't capture `EventManager` state in snapshot, critical in-flight events can be missed. We address this by explicit kicking of events we don't want to miss (virtio queue events). This is safe since virtio queue processing is resilient to spurious events.

_This PR fixes the issue surgically at the Firecracker application level. Long-term solution should fix the issue at the device level or at least at the `Vmm` level._

## Testing

Added regression test that validates emulation is also paused on `VM Pause`.

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

## PR Checklist

`[Author TODO: Meet these criteria.]`
`[Reviewer TODO: Verify that these criteria are met. Request changes if not]`

- [x] All commits in this PR are signed (`git commit -s`).
- [x] The reason for this PR is clearly provided (issue no. or explanation).
- [x] The description of changes is clear and encompassing.
- [x] Any required documentation changes (code and docs) are included in this PR.
- [ ] ~Any newly added `unsafe` code is properly documented.~
- [ ] ~Any API changes are reflected in `firecracker/swagger.yaml`.~
- [ ] ~Any user-facing changes are mentioned in `CHANGELOG.md`.~
- [x] All added/changed functionality is tested.

### CODE REVIEW

File: src/vmm/src/device_manager/persist.rs [hunk 1]
```diff
@@ -223,26 +240,22 @@ impl<'a> Persist<'a> for MMIODeviceManager {
...
+                device.clone(),
+                &net_state.device_id,
+                &net_state.transport_state,
+                &net_state.mmio_slot,
+                constructor_args.event_manager,
+            )?;
+
+            // If device is activated, kick the net queue(s) to make up for any
+            // pending events we may have not captured in snapshot.
+            // Ratelimiters are restored 'unblocked' so nothing to do there.
```
Comment (sandreim) [inline src/vmm/src/device_manager/persist.rs:254]: I guess we can be more explicit here and talk about inflight epoll events.

File: src/devices/src/virtio/net/device.rs [hunk 2]
```diff
@@ -730,6 +730,12 @@ impl Net {
             }
         }
     }
+
+    /// Process device virtio queue(s).
+    pub fn process_virtio_queues(&mut self) {
```
Comment (sandreim) [inline src/devices/src/virtio/net/device.rs:735]: Why no process_virtio_queues() in Vsock ?

File: src/vmm/src/device_manager/persist.rs [hunk 3]
```diff
@@ -162,57 +156,80 @@ impl<'a> Persist<'a> for MMIODeviceManager {
...
-                    BlockConstructorArgs { mem: mem.clone() },
-                    &block_state.device_state,
-                )
-                .map_err(Error::Block)?,
-            ));
 
-            let device_id = block_state.device_id.clone();
-            let transport_state = &block_state.transport_state;
-            let mmio_slot = &block_state.mmio_slot;
+        let mut restore_helper = |device: Arc<Mutex<dyn VirtioDevice>>,
```
Comment (sandreim) [inline src/vmm/src/device_manager/persist.rs:182]: 🙈

**Review by sandreim — COMMENTED** [review]
Looks good, if we cannot reproduce in regression test now, I suggest to create an issue and include it in v0.24.

File: src/vmm/src/device_manager/persist.rs (hunk 1 above)
Comment (acatangiu) [reply src/vmm/src/device_manager/persist.rs]: done

File: src/devices/src/virtio/net/device.rs (hunk 2 above)
Comment (acatangiu) [reply src/devices/src/virtio/net/device.rs:735]: Vsock state machine is not resilient to _control packets_ loss. Epoll events loss is just a small part of the story.

I've opened #2218 to track and discuss this in detail.

I guess for know, we can have it as a _known issue_ with a clear workaround:
> close all vsock connections prior to snapshotting

**Comment by acatangiu** [comment]
@alexandruag @andreeaflorescu @sandreim I've addressed your comments, please take a look!

**Review by sandreim — DISMISSED** [review]

**Review by alexandruag — APPROVED** [review]

**Review by sandreim — APPROVED** [review]

**Pull request merged**